packages/shorebird_cli/lib/src/commands/patch/macos_patcher.dart

```diff
@@ -0,0 +1,446 @@
+        );
+      } catch (error) {
+        buildProgress.fail('$error');
+        rethrow;
+      }
+
+      // Copy the kernel file to the build directory so that it can be used
+      // to generate a patch.
+      macosBuildResult.kernelFile.copySync(_appDillCopyPath);
+
+      buildProgress.complete();
+    } catch (_) {
+      throw ProcessExit(ExitCode.software.code);
+    }
+
+    final appPath = artifactManager.getMacOSAppDirectory()!.path;
+    final tempDir = await Directory.systemTemp.createTemp();
+    final zippedApp = File(p.join(tempDir.path, '${p.basename(appPath)}.zip'));
+    await ditto.archive(
+      source: appPath,
+      destination: zippedApp.path,
+      keepParent: true,
+    );
+    return zippedApp;
+  }
+
+  @override
+  Future<Map<Arch, PatchArtifactBundle>> createPatchArtifacts({
+    required String appId,
+    required int releaseId,
+    required File releaseArtifact,
+    File? supplementArtifact,
+  }) async {
+    // Verify that we have built a patch .app
+    if (artifactManager.getMacOSAppDirectory()?.path == null) {
+      logger.err('Unable to find .app directory');
+      throw ProcessExit(ExitCode.software.code);
+    }
+
+    final unzipProgress = logger.progress('Extracting release artifact');
+    final releaseAppDirectory = Directory.systemTemp.createTempSync();
+    await ditto.extract(
+      source: releaseArtifact.path,
+      destination: releaseAppDirectory.path,
+    );
+
+    File? releaseClassTableLinkInfoFile;
+    File? releaseClassTableLinkDebugInfoFile;
+    if (supplementArtifact != null) {
+      final tempDir = Directory.systemTemp.createTempSync();
+      await artifactManager.extractZip(
+        zipFile: supplementArtifact,
+        outputDirectory: tempDir,
+      );
+      releaseClassTableLinkInfoFile = File(p.join(tempDir.path, 'App.ct.link'));
+      if (!releaseClassTableLinkInfoFile.existsSync()) {
+        logger.err('Unable to find class table link info file');
+        throw ProcessExit(ExitCode.software.code);
+      }
+
+      releaseClassTableLinkDebugInfoFile = File(
+        p.join(tempDir.path, 'App.class_table.json'),
+      );
+      if (!releaseClassTableLinkDebugInfoFile.existsSync()) {
+        logger.err('Unable to find class table link debug info file');
+        throw ProcessExit(ExitCode.software.code);
+      }
+    }
+
+    unzipProgress.complete();
+    final releaseArtifactFile = File(
+      p.join(
+        releaseAppDirectory.path,
+        'Contents',
+        'Frameworks',
+        'App.framework',
+        'App',
+      ),
+    );
+
+    final useLinker = AotTools.usesLinker(shorebirdEnv.flutterRevision);
+    if (useLinker) {
+      // If we're using a newer version of the linker, we need to also copy the
+      // necessary class table link information alongside the snapshots.
+      if (releaseClassTableLinkInfoFile != null &&
+          releaseClassTableLinkDebugInfoFile != null) {
```

Comment:
These should never be null so we should refactor this to exit early if either one is null on MacOS